Create Cache as a Customizable Object

mrambacher opened the pull request:
This is a predecessor to #8998.  This change adds the ability to create a Cache from the ObjectRegistry.  Unlike 8998, each call to Cache::CreateFromString will create a new object -- no managed object code is involved here.

pdillinger commented on cache/sharded_cache.h:
```diff
@@ -111,19 +108,29 @@ class ShardedCache : public Cache {
+  uint32_t SetNumShards(int num_shard_bits);
+
   inline uint32_t Shard(uint32_t hash) { return hash & shard_mask_; }
+  mutable port::Mutex capacity_mutex_;
```
I would call this `options_mutex_` and comment that it is used to ensure consistency between the "source of truth" `options_` in each derived class and the downstream option settings in each cache shard. This "source of truth" is not normally read during performance-critical operations, so a Cache-wide mutex is OK.

pdillinger replied:
Previously it was only used for capacity-related mutability. Now it's clearly generalized to guarding all dynamic options, based on new usage in `IsMutable` and `PrepareOptions`

mrambacher replied:
Done.

pdillinger commented on include/rocksdb/cache.h:
```diff
@@ -535,9 +537,6 @@ class Cache {
   virtual void WaitAll(std::vector<Handle*>& /*handles*/) {}
-
- private:
-  std::shared_ptr<MemoryAllocator> memory_allocator_;
```
OK by me. As long as we are clear in HISTORY.md that the API updates to Cache require some updates to custom implementations, I think it will be fine.

pdillinger commented on options/customizable.cc:
```diff
@@ -70,6 +70,25 @@ std::string Customizable::SerializeOptions(const ConfigOptions& config_options,
+  } else {
+    auto ipos = id.find("@");
+    auto tpos = that_id.find("@");
+    if (ipos != std::string::npos && tpos != std::string::npos) {
```
Here or in API comment say that it ignores trailing `@<anything>`?

pdillinger reviewed (changes requested):
* Rebase or merge to fix protobuf CI issue
* Fix clang-analyze issue
* Add to HISTORY.md
* Other inline suggestions

pdillinger reviewed (changes requested):
Rename and comment on mutex

mrambacher commented:
@pdillinger I believe I resolved your comments.  As for the "analyze" test that is failing, I tried on an ubuntu docker container with a comparable setup (same version of clang, etc) and see no errors.  Would there be any way to get the errors off the Circle box into the log or some other way to tell what is failing?  TIA!

pdillinger commented:
The errors are in the output (with enough scrolling)

```
cache/cache_bench_tool.cc:809:35: warning: Use of zero-allocated memory
      dbs_[db_i].orig_file_number += 1;
                                  ^
cache/cache_bench_tool.cc:900:32: warning: Use of zero-allocated memory
      dbs_[0].orig_file_number = 0;
                               ^
2 warnings generated.
```

pdillinger commented:
Weird error. I thought it was talking about zero-initialized memory, but I think it means `db_count` might be zero. If you coerce or assert `db_count >= 1` that should suffice.

pdillinger reviewed (approved):
Seems to be good now

jay-zhuang commented on cache/clock_cache.cc:
```diff
@@ -766,74 +759,259 @@ void ClockCacheShard::EraseUnRefEntries() {
-      shards_ = nullptr;
+Status ClockCache::PrepareOptions(const ConfigOptions& config_options) {
+  Status s;
+  if (!IsClockCacheSupported()) {
+    (void)config_options;
+    s = Status::NotSupported("ClockCache not supported");
+#ifdef SUPPORT_CLOCK_CACHE
```
nit:
move this to `#else // SUPPORT_CLOCK_CACHE`:
```
#else
    (void)config_options;
    s = Status::NotSupported("ClockCache not supported");
#endif // SUPPORT_CLOCK_CACHE
```

jay-zhuang commented on cache/clock_cache.h:
```diff
@@ -9,8 +9,93 @@
+class ClockCacheShard;
+
+struct ClockCacheOptions {
+  ClockCacheOptions() {}
```
nit:
```suggestion
  ClockCacheOptions() = default;
```

jay-zhuang commented on cache/clock_cache.h:
```diff
@@ -9,8 +9,93 @@
+  const char* Name() const override { return kClassName(); }
+  Status PrepareOptions(const ConfigOptions& options) override;
+  bool IsMutable() const override;
+  virtual std::string GetPrintableOptions() const override;
+
+  virtual void SetCapacity(size_t capacity) override;
+  virtual void SetStrictCapacityLimit(bool strict_capacity_limit) override;
+  virtual size_t GetCapacity() const override;
+  virtual bool HasStrictCapacityLimit() const override;
```
nit, virtual is redundant with override:
```suggestion
  std::string GetPrintableOptions() const override;

  void SetCapacity(size_t capacity) override;
  void SetStrictCapacityLimit(bool strict_capacity_limit) override;
  size_t GetCapacity() const override;
  bool HasStrictCapacityLimit() const override;
```

jay-zhuang commented on options/customizable.cc:
```diff
@@ -15,6 +15,22 @@
+namespace {
+// Splits an IndividualId into its three components (name@addr#pid).
+// Returns true if the input id had all components and false otherwise.
+static bool IsIndividualId(const std::string& id, size_t* addr,
```
nit:
```suggestion
bool IsIndividualId(const std::string& id, size_t* addr,
```

jay-zhuang commented on options/customizable.cc:
```diff
@@ -15,6 +15,22 @@
+static bool IsIndividualId(const std::string& id, size_t* addr,
+                           size_t* pid = nullptr) {
+  auto hoff = std::string::npos;
+  *addr = id.find("@");
```
nit:
```suggestion
  *addr = id.find('@');
```

jay-zhuang commented on options/customizable.cc:
```diff
@@ -15,6 +15,22 @@
+  auto hoff = std::string::npos;
+  *addr = id.find("@");
+  if (*addr != std::string::npos) {
+    hoff = id.find("#", *addr);
```
```suggestion
    hoff = id.find('#', *addr);
```

jay-zhuang reviewed (approved):
LGTM
A few minor issues would be nice to fix.

mrambacher commented:
@pdillinger @jay-zhuang I think I have addressed all of the PR comments.  Can this be merged soon?  Keeping up with the Cache changes is getting to be tricky.  Thanks!

pdillinger commented:
May have chosen a bad time to merge 🙁 (#10434 )

pdillinger commented:
Also, we are concerned about the race conditions in customizable / setoptions infrastructure, e.g. #10079 and #10375. Plus we're cautious about other functional regressions like the one patched in #10378

mrambacher commented:
> Also, we are concerned about the race conditions in customizable / setoptions infrastructure, e.g. #10079 and #10375. Plus we're cautious about other functional regressions like the one patched in #10378

I created #10387 to eliminate many of the extra calls to PrepareOptions that were potential with #10375.   In #9756, I changed code that will also eliminate some of the potential PrepareOptions calls (if bad name=value were passed to Configure), as well as improve the performance.  

I do not know if these PRs will address the other race conditions, but I would think they might and are worth a look.

Also note that if Cache is not marked as Mutable, it cannot be changed via SetOptions.  If the properties themselves (e.g. "capacity" are not mutable, they cannot be changed via SetOptions and only via the corresponding API.  I doubt the same race condition(s) apply here.
